staging/src/k8s.io/apiextensions-apiserver/test/integration/finalization_test.go

```diff
@@ -89,3 +92,77 @@ func TestFinalization(t *testing.T) {
 	require.Error(t, err)
 	require.True(t, errors.IsNotFound(err), "%#v", err)
 }
+
+func TestFinalizationAndDeletion(t *testing.T) {
+	stopCh, apiExtensionClient, clientPool, err := testserver.StartDefaultServerWithClients()
+	require.NoError(t, err)
+	defer close(stopCh)
+
+	// Create a CRD.
+	noxuDefinition := testserver.NewNoxuCustomResourceDefinition(apiextensionsv1beta1.ClusterScoped)
+	noxuVersionClient, err := testserver.CreateNewCustomResourceDefinition(noxuDefinition, apiExtensionClient, clientPool)
+	require.NoError(t, err)
+
+	// Create a CR with a finalizer.
+	ns := "not-the-default"
+	name := "foo123"
+	noxuResourceClient := NewNamespacedCustomResourceClient(ns, noxuVersionClient, noxuDefinition)
+
+	instance := testserver.NewNoxuInstance(ns, name)
+	instance.SetFinalizers([]string{"noxu.example.com/finalizer"})
+	createdNoxuInstance, err := instantiateCustomResource(t, instance, noxuResourceClient, noxuDefinition)
+	require.NoError(t, err)
+
+	// Delete a CR. Because there's a finalizer, it will not get deleted now.
+	uid := createdNoxuInstance.GetUID()
+	err = noxuResourceClient.Delete(name, &metav1.DeleteOptions{
+		Preconditions: &metav1.Preconditions{
+			UID: &uid,
+		},
+	})
+	require.NoError(t, err)
+
+	// Check is the CR scheduled for deletion.
+	gottenNoxuInstance, err := noxuResourceClient.Get(name, metav1.GetOptions{})
+	require.NoError(t, err)
+	require.NotNil(t, gottenNoxuInstance.GetDeletionTimestamp())
+
+	// Delete the CRD.
+	testserver.DeleteCustomResourceDefinition(noxuDefinition, apiExtensionClient)
+
+	// Check is CR still there after the CRD deletion.
+	gottenNoxuInstance, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+	require.NoError(t, err)
+
+	// Update the CR to remove the finalizer.
+	for {
+		gottenNoxuInstance.SetFinalizers(nil)
+		_, err = noxuResourceClient.Update(gottenNoxuInstance)
+		if err == nil {
+			break
+		}
+		if !errors.IsConflict(err) {
+			require.NoError(t, err) // Fail on unexpected error
+		}
+		gottenNoxuInstance, err = noxuResourceClient.Get(name, metav1.GetOptions{})
+		require.NoError(t, err)
+	}
+
+	// Verify the CR is gone.
+	// It should return the NonFound error.
+	err = wait.Poll(500*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
```

Comment:
Why does it need a Poll here? If the finalizer is removed, the CR should be deleted immediately.

Reply:
As far as I understood, it can take few seconds sometimes to get delete. [More details](https://github.com/kubernetes/kubernetes/pull/60592#discussion_r171468946).

Reply:
@caesarxuchao do we guarantee that? From the Finalizer field godoc I only ready that it might be deleted. I don't see that it guaranteed to be deleted right away synchronously in the UPDATE request. Moreover, HA servers might still see it from their caches.

Reply:
@caesarxuchao Does this lgty?

Reply:
If the Update removes the last finalizer, the generic registry deletes the object synchronously (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L631). The CR registry inherits the generic registry, so that's the expected behavior of CR as well. Also, this particular test isn't using the HA setup, so I think we don't need the poll.

If you agree, we can send a followup PR to update field godoc.

Reply:
@caesarxuchao @nikhita It seems like the test is passing even without that poll. Should I remove it from the test?

Reply:
Yes, please remove. Then lgtm.

Reply:
@caesarxuchao @nikhita Done :)